### ISSUE TITLE
AutoDiffCostFunction supports only double floating point numbers

### ISSUE THREAD

**Issue by csukuangfj** [ceres-solver/ceres-solver#421]
See the code
https://github.com/ceres-solver/ceres-solver/blob/72d8e1cdb1e7cdfbd662892f2a04f9af792a9be7/include/ceres/autodiff_cost_function.h#L202-L204

though `AutoDiff` supports any type `T`
https://github.com/ceres-solver/ceres-solver/blob/72d8e1cdb1e7cdfbd662892f2a04f9af792a9be7/include/ceres/internal/autodiff.h#L205-L208

It is fixed to `double` in `AutoDiffCostFunction` which forbids users using `float`.

**Comment by sandwichmaker** [comment]
Yes this is by design, since Ceres only support double  valued cost functions.

**Comment by slitcch** [comment]
Why does Ceres only support double-valued cost functions? I have to make my thing run really really fast, and I suspect that single-precision will work fine with my problem (autodiff and pretty sparse)

**Comment by slitcch** [comment]
If I wanted to hack single-precision in, could you tell me roughly where I should be changing things? I'm (probably) happy to take it on even if it's a lot.

**Comment by slitcch** [comment]
Thanks for responding @sandwichmaker ! I actually ended up going with `TinySolver`, which I soon after realized supports single floats natively. It works really well - for my very sparse application (~42-150 residuals) it's a lot faster than the normal Solver! And using single-precision is around 30% faster than double. Overall everything is very very fast now; might take another speed optimization pass in a few months, but for now I have no complaints :)

I also packaged a very small derivative of Ceres with just TinySolver and dependencies [here](https://gitlab.freedesktop.org/monado/utilities/hand-tracking-playground/tinyceres) just in case that is of interest to anyone :)
